dangerzone/updater/signatures.py

```diff
@@ -0,0 +1,626 @@
+
+__all__ = [
+    "verify_signature",
+    "load_and_verify_signatures",
+    "store_signatures",
+    "verify_offline_image_signature",
+]
+
+
+def signature_to_bundle(sig: Dict) -> Dict:
+    """Convert a cosign-download signature to the format expected by cosign bundle."""
+    bundle = sig["Bundle"]
+    payload = bundle["Payload"]
+    return {
+        "base64Signature": sig["Base64Signature"],
+        "Payload": sig["Payload"],
+        "cert": sig["Cert"],
+        "chain": sig["Chain"],
+        "rekorBundle": {
+            "SignedEntryTimestamp": bundle["SignedEntryTimestamp"],
+            "Payload": {
+                "body": payload["body"],
+                "integratedTime": payload["integratedTime"],
+                "logIndex": payload["logIndex"],
+                "logID": payload["logID"],
+            },
+        },
+        "RFC3161Timestamp": sig["RFC3161Timestamp"],
+    }
+
+
+def verify_signature(signature: dict, image_digest: str, pubkey: Path) -> None:
+    """
+    Verifies that the signature matches the given public key and image digest.
+    """
+    # FIXME Also verify the identity/docker-reference field against
+    # `container_utils.expected_image_name()`
+    # e.g. ghcr.io/freedomofpress/dangerzone/dangerzone
+
+    signature_bundle = signature_to_bundle(signature)
+    try:
+        payload_bytes = b64decode(signature_bundle["Payload"])
+        payload_digest = json.loads(payload_bytes)["critical"]["image"][
+            "docker-manifest-digest"
+        ]
+    except Exception as e:
+        raise errors.SignatureVerificationError(
+            f"Unable to extract the payload digest from the signature: {e}"
+        )
+    if payload_digest != f"sha256:{image_digest}":
+        raise errors.SignatureMismatch(
+            "The given signature does not match the expected image digest "
+            f"({payload_digest}, {image_digest})"
+        )
+
+    # Note: Pass delete=False here to avoid deleting the file before usage.
+    # The file is read outside of the context manager otherwise it fails
+    # on Windows, where the file can't be opened concurrently.
+    #
+    # Using a O_TEMPORARY flag mentioned in [0] is not practical here as how
+    # cosign opens the file is not configurable.
+    #
+    # Hence the os.remove calls after calling cosign.verify_blob.
+    #
+    # [0] https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile
+    with (
+        NamedTemporaryFile(mode="w", delete=False) as signature_file,
+        NamedTemporaryFile(mode="bw", delete=False) as payload_file,
+    ):
+        json.dump(signature_bundle, signature_file)
+        signature_file.flush()
+
+        payload_file.write(payload_bytes)
+        payload_file.flush()
+
+    cosign.verify_blob(pubkey, signature_file.name, payload_file.name)
+    log.debug("Signature verified")
+
+    os.remove(signature_file.name)
+    os.remove(payload_file.name)
+
+
+@dataclass
+class Signature:
```

Comment:
This class is mainly used in the tests, I wonder if we should try to use it in this module?